interpreter/valid/valid.ml

```diff
@@ -201,9 +201,15 @@ let check_globaltype (c : context) (gt : globaltype) at =
 let check_memorytype (c : context) (mt : memorytype) at =
   let MemoryT (at_, lim, pt) = mt in
   let sz, s =
-    match at_ with
-    | I32AT -> 0x1_0000L, "2^16 pages (4 GiB) for i32"
-    | I64AT -> 0x1_0000_0000_0000L, "2^48 pages (256 TiB) for i64"
+    match pt with
+    | PageT 0x10000 ->
+       (match at_ with
+       | I32AT -> 0x1_0000L, "2^16 pages (4 GiB) for i32"
+       | I64AT -> 0x1_0000_0000_0000L, "2^48 pages (256 TiB) for i64")
+    | _ -> (* TODO: divide by page size, what about error msg? *)
+       (match at_ with
+       | I32AT -> 0xFFFF_FFFFL, "2^32 - 1 bytes for i32"
+       | I64AT -> 0xFFFF_FFFF_FFFF_FFFFL, "2^64 - 1 bytes for i64")
   in
   check_limits lim sz at ("memory size must be at most " ^ s);
   check_pagetype pt at
```

Comment:
Perhaps move this up, so that it's validated before we're matching on it to validate other parts.